homeassistant/components/ambient_station/__init__.py

```diff
@@ -0,0 +1,214 @@
+            vol.Optional(CONF_UNIT_SYSTEM):
+                vol.In([UNITS_SI, UNITS_US]),
+        })
+}, extra=vol.ALLOW_EXTRA)
+
+
+async def async_setup(hass, config):
+    """Set up the Ambient PWS component."""
+    hass.data[DOMAIN] = {}
+    hass.data[DOMAIN][DATA_CLIENT] = {}
+
+    if DOMAIN not in config:
+        return True
+
+    conf = config[DOMAIN]
+
+    if conf[CONF_APP_KEY] in configured_instances(hass):
+        return True
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={'source': SOURCE_IMPORT}, data=conf))
+
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
+    """Set up the Ambient PWS as config entry."""
+    from aioambient import Client
+    from aioambient.errors import WebsocketConnectionError
+
+    session = aiohttp_client.async_get_clientsession(hass)
+
+    try:
+        ambient = AmbientStation(
+            hass,
+            config_entry,
+            Client(
+                config_entry.data[CONF_API_KEY],
+                config_entry.data[CONF_APP_KEY], session),
+            config_entry.data.get(
+                CONF_MONITORED_CONDITIONS, list(SENSOR_TYPES)),
+            config_entry.data.get(CONF_UNIT_SYSTEM))
+        hass.async_create_task(ambient.ws_connect())
+        hass.data[DOMAIN][DATA_CLIENT][config_entry.entry_id] = ambient
+    except WebsocketConnectionError as err:
+        _LOGGER.error('Config entry failed: %s', err)
+        raise ConfigEntryNotReady
+
+    return True
+
+
+async def async_unload_entry(hass, config_entry):
+    """Unload an OpenUV config entry."""
+    ambient = hass.data[DOMAIN][DATA_CLIENT].pop(config_entry.entry_id)
+    hass.async_create_task(ambient.ws_disconnect())
+
+    await hass.config_entries.async_forward_entry_unload(
+        config_entry, 'sensor')
+
+    return True
+
+
+class AmbientStation:
+    """Define a class to handle the Ambient websocket."""
+
+    def __init__(
+            self, hass, config_entry, client, monitored_conditions,
+            unit_system):
+        """Initialize."""
+        self._client = client
+        self._config_entry = config_entry
+        self._hass = hass
+        self._lock = RLock()
+        self._ws_reconnect_handle = None
+        self.monitored_conditions = monitored_conditions
+        self.stations = {}
+        self.unit_system = unit_system
+
+        hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_STOP, client.websocket.disconnect())
+
+    async def ws_connect(self):
+        """Register handlers and connect to the websocket."""
+        from aioambient.errors import WebsocketError
+
+        def on_connect():
+            """Define a handler to fire when the websocket is connected."""
+            _LOGGER.info('Connected to websocket')
+
+        def on_data(data):
+            """Define a handler to fire when the data is received."""
+            with self._lock:
```

Comment:
Why do we need the lock? This is a threading lock. Why do we use that in async context?

Reply:
Great point. 😆